File: C#/Silverlight/source/LSilverlight-0.3.3/Loon.Action/ActionEvent.cs
namespace Loon.Action {

    using Loon.Core.Timer;

	public abstract class ActionEvent {
	
		private LTimer timer;
	
		private ActionListener actionListener;
	
		internal bool firstTick, isComplete, isInit;

        internal Event original;
	
		internal object tag;
	
		internal float offsetX, offsetY;
	
		public ActionEvent() {
			timer = new LTimer(0);
		}
	
		public long GetDelay() {
			return timer.GetDelay();
		}
	
		public void SetDelay(long d) {
			timer.SetDelay(d);
		}
	
		public void Paused(bool pause) {
			ActionControl.GetInstance().Paused(pause, original);
		}
	
		public void Step(long elapsedTime) {
			if (original == null) {
				return;
			}
			if (timer.Action(elapsedTime)) {
				if (firstTick) {
					this.firstTick = false;
					this.timer.Refresh();
				} else {
					Update(elapsedTime);
				}
				if (actionListener != null) {
					actionListener.Process(original);
				}
			}
		}
	
		public object GetOriginal() {
			return original;
		}

        public virtual void Start(Event o)
        {
			if (o == null) {
				return;
			}
			this.original = o;
			this.timer.Refresh();
			this.firstTick = true;
			this.isComplete = false;
			this.isInit = false;
			if (actionListener != null) {
				actionListener.Start(o);
			}
		}
	
		public abstract void Update(long elapsedTime);
	
		public abstract void OnLoad();
	
		public void Stop() {
			if (actionListener != null) {
				actionListener.Stop(original);
			}
		}
	
		public abstract bool IsComplete();
	
		public object GetTag() {
			return tag;
		}
	
		public void SetTag(object tag_0) {
			this.tag = tag_0;
		}
	
		public void SetComplete(bool isComplete_0) {
			this.isComplete = isComplete_0;
		}
	
		public ActionListener GetActionListener() {
			return actionListener;
		}
	
		public void SetActionListener(ActionListener actionListener_0) {
			this.actionListener = actionListener_0;
		}
	
		public void SetOffset(float x, float y) {
			this.offsetX = x;
			this.offsetY = y;
		}
	
		public float GetOffsetX() {
			return offsetX;
		}
	
		public void SetOffsetX(float offsetX_0) {
			this.offsetX = offsetX_0;
		}
	
		public float GetOffsetY() {
			return offsetY;
		}
	
		public void SetOffsetY(float offsetY_0) {
			this.offsetY = offsetY_0;
		}
	}
}

File: Java/Canvas_ver/src/org/loon/framework/android/game/action/ActionEvent.java
package org.loon.framework.android.game.action;

import org.loon.framework.android.game.core.graphics.component.Actor;
import org.loon.framework.android.game.core.timer.LTimer;

/**
 * Copyright 2008 - 2011
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loonframework
 * @author chenpeng
 * @email：ceponline@yahoo.com.cn
 * @version 0.1
 */
public abstract class ActionEvent {

	private LTimer timer;

	private ActionListener actionListener;

	boolean firstTick, isComplete, isInit;

	Actor original;

	Object tag;

	public ActionEvent() {
		timer = new LTimer(0);
	}

	public long getDelay() {
		return timer.getDelay();
	}

	public void setDelay(long d) {
		timer.setDelay(d);
	}

	public void step(long elapsedTime) {
		if (original == null) {
			return;
		}
		if (firstTick) {
			this.firstTick = false;
			this.timer.refresh();
		} else {
			if (timer.action(elapsedTime)) {
				update(elapsedTime);
			}
		}
		if (actionListener != null) {
			actionListener.process(original);
		}
	}

	public Object getOriginal() {
		return original;
	}

	public void start(Actor o) {
		if (o == null) {
			return;
		}
		this.original = o;
		this.timer.refresh();
		this.firstTick = true;
		this.isComplete = false;
		this.isInit = false;
		if (actionListener != null) {
			actionListener.start(o);
		}
	}

	public abstract void update(long elapsedTime);

	public abstract void onLoad();

	public void stop() {
		if (actionListener != null) {
			actionListener.stop(original);
		}
	}

	public abstract boolean isComplete();

	public Object getTag() {
		return tag;
	}

	public void setTag(Object tag) {
		this.tag = tag;
	}

	public final void setComplete(boolean isComplete) {
		this.isComplete = isComplete;
	}

	public ActionListener getActionListener() {
		return actionListener;
	}

	public void setActionListener(ActionListener actionListener) {
		this.actionListener = actionListener;
	}
}
